File: js/src/resources/edit-image.ts
import type { HttpClient, RequestOptions, PollingOptions, ActionSchema } from '@runapi.ai/core';
import { compactParams, validateParams } from '@runapi.ai/core';
import { pollUntilComplete } from '@runapi.ai/core/internal';
import { contract } from '../contract_gen';
import type {
  CompletedEditImageResponse,
  EditImageParams,
  EditImageResponse,
  TaskCreateResponse,
} from '../types';

const ENDPOINT = '/api/v1/qwen_image/edit_image';

/** Applies targeted edits to a source image using natural-language prompts. */
export class EditImage {
  constructor(private readonly http: HttpClient) {}

  /**
   * Edit an image with a natural-language prompt and wait until complete.
   * @param params Edit-image parameters.
   * @param options Per-request and polling overrides.
   * @returns The completed edit-image result.
   */
  async run(params: EditImageParams, options?: RequestOptions & PollingOptions): Promise<CompletedEditImageResponse> {
    const { id } = await this.create(params, options);
    const response = await pollUntilComplete<EditImageResponse>(() => this.get(id, options), {
      maxWaitMs: options?.maxWaitMs,
      pollIntervalMs: options?.pollIntervalMs,
    });
    return response as CompletedEditImageResponse;
  }

  /**
   * Create an edit-image task; returns immediately with a task id.
   * @param params Edit-image parameters.
   * @param options Per-request overrides.
   * @returns The task creation result with id.
   */
  async create(params: EditImageParams, options?: RequestOptions): Promise<TaskCreateResponse> {
    const body = compactParams(params);
    validateParams(contract['edit-image'] as ActionSchema, body as Record<string, unknown>);
    return this.http.request<TaskCreateResponse>('POST', ENDPOINT, {
      body,
      ...options,
    });
  }

  /**
   * Fetch the current status of an edit-image task.
   * @param id The task id.
   * @param options Per-request overrides.
   * @returns The current edit-image task status.
   */
  async get(id: string, options?: RequestOptions): Promise<EditImageResponse> {
    return this.http.request<EditImageResponse>('GET', `${ENDPOINT}/${id}`, {
      ...options,
    });
  }
}

File: ruby/lib/runapi/qwen_image/resources/edit_image.rb
# frozen_string_literal: true

module RunApi
  module QwenImage
    module Resources
      # QwenImage edit-image resource. Edits images with natural-language prompts.
      class EditImage
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/qwen_image/edit_image"

        RESPONSE_CLASS = Types::EditImageResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedEditImageResponse

        def initialize(http)
          @http = http
        end

        # Edit an image and wait until complete.
        #
        # @param params [Hash] edit-image parameters
        # @return [RunApi::QwenImage::Types::CompletedEditImageResponse] completed edit-image result
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Create an edit-image task.
        #
        # @param params [Hash] edit-image parameters
        # @return [RunApi::QwenImage::Types::EditImageResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_contract!(CONTRACT["edit-image"], params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get edit-image status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::QwenImage::Types::EditImageResponse] current edit-image status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end
      end
    end
  end
end
